ssl/ssl_sess.c

```diff
@@ -815,6 +815,9 @@ static int remove_session_lock(SSL_CTX *ctx, SSL_SESSION *c, int lck)
             if (ctx->remove_session_cb != NULL)
                 ctx->remove_session_cb(ctx, r);
             SSL_SESSION_free(r);
+        } else {
+            if (ctx->remove_session_cb != NULL)
+                ctx->remove_session_cb(ctx, c);
```

Comment:
Am I missing something or is it not the case that `r == c` in which case this code could be simplified to always do the `remove_session_cb` stuff regardless of the value of `ret`. Also I wonder whether `not_resumable` shouldn't also be set in both cases, i.e. the only thing that depends on `ret` is the `SSL_SESSION_free()` call.